server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -104,6 +118,63 @@ protected void shardOperationOnPrimary(
+        final ShardId sourceShard = request.shardId();
+        IndexMetadata indexMetadata = project.getIndexSafe(request.shardId().getIndex());
+        SplitShardCountSummary shardCountSummary = SplitShardCountSummary.forIndexing(indexMetadata, sourceShard.getId());
+        Map<ShardId, BasicReplicationRequest> requestsByShard = new HashMap<>();
+        requestsByShard.put(sourceShard, request);
+        // Create a request for original source shard and for each target shard.
+        // New requests that are to be handled by target shards should contain the
+        // latest ShardCountSummary.
+        int targetShardId = indexMetadata.getReshardingMetadata().getSplit().targetShard(sourceShard.id());
+        ShardId targetShard = new ShardId(request.shardId().getIndex(), targetShardId);
+        requestsByShard.put(targetShard, new BasicReplicationRequest(targetShard, shardCountSummary));
+        return requestsByShard;
+    }
+
+    @Override
+    protected Tuple<ReplicationResponse, Exception> combineSplitResponses(
+        BasicReplicationRequest originalRequest,
+        Map<ShardId, BasicReplicationRequest> splitRequests,
+        Map<ShardId, Tuple<ReplicationResponse, Exception>> responses
+    ) {
+        int failed = 0;
+        int successful = 0;
+        int total = 0;
+        List<ReplicationResponse.ShardInfo.Failure> failures = new ArrayList<>();
+
+        // If the action fails on either one of the shards, we return an exception.
+        // Case 1: Both source and target shards return a response: Add up total, successful, failures
+        // Case 2: Both source and target shards return an exception : return exception
+        // Case 3: One shards returns a response, the other returns an exception : return exception
+        for (Map.Entry<ShardId, Tuple<ReplicationResponse, Exception>> entry : responses.entrySet()) {
+            ShardId shardId = entry.getKey();
+            Tuple<ReplicationResponse, Exception> value = entry.getValue();
+            Exception exception = value.v2();
+            if (exception != null) {
+                return new Tuple<>(null, exception);
+            } else {
+                ReplicationResponse response = value.v1();
+                failed += response.getShardInfo().getFailed();
+                successful += response.getShardInfo().getSuccessful();
+                total += response.getShardInfo().getTotal();
+                Collections.addAll(failures, response.getShardInfo().getFailures());
+            }
+        }
+        ReplicationResponse.ShardInfo.Failure[] failureArray = failures.toArray(new ReplicationResponse.ShardInfo.Failure[0]);
+        assert failureArray.length == failed;
+        ReplicationResponse.ShardInfo shardInfo = ReplicationResponse.ShardInfo.of(total, successful, failureArray);
+        ReplicationResponse response = new ReplicationResponse();
+        response.setShardInfo(shardInfo);
+        return new Tuple<>(response, null);
+    }
```

Comment:
Can we share this logic with the flush action instead of duplicating it? That way e.g. fixing the possible stale cluster read happens in one place.